### PR TITLE
[clusteragent/autoscaling] Check that autoscaling target is not cluster agent

### PR DESCRIPTION
**Pull request by jennchenn**
### What does this PR do?

Check that the autoscaling target is not set to the cluster agent. If it is, set an error on the corresponding `DatadogPodAutoscaler`.

### Motivation

We do not want the cluster agent to be taken down due to autoscaler failures.

### Additional Notes

If `CLUSTER_AGENT_DEPLOYMENT` is not set, it defaults to assuming `datadog-agent-cluster-agent` as the deployment name.

Do we want to add `CLUSTER_AGENT_DEPLOYMENT` to the agent config rather than reading the environment variable directly? 

### Describe how to test/QA your changes

1. Create a `DatadogPodAutoscaler`resource that targets the cluster agent
2. Verify that an error status is set and the cluster agent is not scaled by it
```
❯ k describe datadogpodautoscaler cluster-agent

...
Status:
  Conditions:
    Last Transition Time:  2024-08-23T21:17:55Z
    Reason:                Autoscaling target cannot be set to the cluster agent
    Status:                True
    Type:                  Error
    Last Transition Time:  2024-08-23T21:17:55Z
...
```

### CODE REVIEW

File: pkg/clusteragent/autoscaling/workload/controller.go [hunk 1]
```diff
@@ -389,3 +408,10 @@ func (c *Controller) deletePodAutoscaler(ns, name string) error {
 	}
 	return nil
 }
+
+func getDeploymentName() string {
```
Comment (vboulineau) [inline pkg/clusteragent/autoscaling/workload/controller.go:412]: Tooling already exists for that: https://github.com/DataDog/datadog-agent/blob/main/pkg/util/kubernetes/apiserver/common/common_linux.go

It reads `DD_POD_NAME` which is already set and defaults to `os.hostname()` (should work for DCA). From there you can use a variant of https://github.com/DataDog/datadog-agent/blob/main/pkg/util/kubernetes/helpers.go#L22-L40 to get Deployment name from POD Name

File: pkg/clusteragent/autoscaling/workload/controller.go [hunk 2]
```diff
@@ -272,6 +274,23 @@ func (c *Controller) syncPodAutoscaler(ctx context.Context, key, ns, name string
 	// Reaching this point, we had an error in processing, clearing up global error
 	podAutoscalerInternal.SetError(nil)
 
+	// Check that targetRef is not set to the cluster agent
+	clusterAgentRef := autoscalingv2.CrossVersionObjectReference{
```
Comment (vboulineau) [inline pkg/clusteragent/autoscaling/workload/controller.go:278]: You also need to check autoscaler namespace vs cluster agent namespace: https://github.com/DataDog/datadog-agent/blob/main/pkg/util/kubernetes/apiserver/common/common.go#L40-L49

Also not sure you need to compare the full `CrossVersionObjectReference`. Basically if something has some name and namespace than DCA it's suspicious enough.

File: pkg/clusteragent/autoscaling/workload/controller.go [hunk 3]
```diff
@@ -272,6 +274,23 @@ func (c *Controller) syncPodAutoscaler(ctx context.Context, key, ns, name string
...
+		Kind:       "Deployment",
+		Name:       getDeploymentName(),
+		APIVersion: "apps/v1",
+	}
+
+	if podAutoscaler.Spec.TargetRef == clusterAgentRef {
```
Comment (vboulineau) [inline pkg/clusteragent/autoscaling/workload/controller.go:284]: nit: The code should probably go to a dedicated function like `validateAutoscaler` or `validateRequirements`, the max 100 could also go there.

File: pkg/clusteragent/autoscaling/workload/controller.go (hunk 1 above)
Comment (jennchenn) [reply pkg/clusteragent/autoscaling/workload/controller.go]: Thanks! Made a new helper to get the deployment name from the pod name which reuses the code from the helper you shared here

File: pkg/clusteragent/autoscaling/workload/controller.go (hunk 2 above)
Comment (vboulineau) [reply pkg/clusteragent/autoscaling/workload/controller.go]: Thinking more about it, you should try to cover both Cluster Agent Deployment AND direct ReplicaSet

File: pkg/clusteragent/autoscaling/workload/controller.go [hunk 4]
```diff
@@ -389,3 +388,35 @@ func (c *Controller) deletePodAutoscaler(ns, name string) error {
...
+	deploymentName := kubernetes.ParseDeploymentForPodName(clusterAgentPodName)
+	clusterAgentNs := common.GetMyNamespace()
+
+	if podAutoscaler.Namespace == clusterAgentNs && podAutoscaler.Spec.TargetRef.Name == deploymentName {
+		return fmt.Errorf("Autoscaling target cannot be set to the cluster agent")
+	}
+	return nil
+}
+
+func (c *Controller) updateAutoscalerStatus(ctx context.Context, key, ns, name string, err error, podAutoscalerInternal model.PodAutoscalerInternal, podAutoscaler *datadoghq.DatadogPodAutoscaler) error {
```
Comment (vboulineau) [inline pkg/clusteragent/autoscaling/workload/controller.go:408]: I'd expect the method to highlight that it unlocks store and one line it:
```
		return autoscaling.NoRequeue, c.updateAndUnlock(ctx, key, ns, name, validationErr, podAutoscalerInternal, podAutoscaler)
```

There's probably a better name to find for it!

**Review by sblumenthal — APPROVED** [review]

**Review by davidor — APPROVED** [review]
:+1: for container-platform files

**Review by mfpierre — APPROVED** [review]

**Pull request merged**